pkg/cloudprovider/providers/rackspace/rackspace.go

```diff
@@ -391,3 +467,145 @@ func (os *Rackspace) GetZone() (cloudprovider.Zone, error) {
+			glog.V(4).Infof("Disk: %q is already attached to compute: %q", diskName, instanceID)
+			return disk.ID, nil
+		}
+
+		errMsg := fmt.Sprintf("Disk %q is attached to a different compute: %q, should be detached before proceeding", diskName, disk.Attachments[0]["server_id"])
+		glog.Errorf(errMsg)
+		return "", errors.New(errMsg)
+	}
+
+	_, err = volumeattach.Create(compute, instanceID, &osvolumeattach.CreateOpts{
+		VolumeID: disk.ID,
+	}).Extract()
+	if err != nil {
+		glog.Errorf("Failed to attach %s volume to %s compute", diskName, instanceID)
+		return "", err
+	}
+	glog.V(2).Infof("Successfully attached %s volume to %s compute", diskName, instanceID)
+	return disk.ID, nil
+}
+
+func (rs *Rackspace) GetDevicePath(diskId string) string {
+	volume, err := rs.getVolume(diskId)
+	if err != nil {
+		return ""
+	}
+	attachments := volume.Attachments
+	if len(attachments) != 1 {
+		glog.Warningf("Unexpected number of volume attachements on %s: %d", diskId, len(attachments))
+		return ""
+	}
+	return attachments[0]["device"].(string)
+}
+
+// Takes a partial/full disk id or diskname
+func (rs *Rackspace) getVolume(diskName string) (volumes.Volume, error) {
+	sClient, err := rackspace.NewBlockStorageV1(rs.provider, gophercloud.EndpointOpts{
+		Region: rs.region,
+	})
+
+	var volume volumes.Volume
+	if err != nil || sClient == nil {
+		glog.Errorf("Unable to initialize cinder client for region: %s", rs.region)
+		return volume, err
+	}
+
+	err = volumes.List(sClient).EachPage(func(page pagination.Page) (bool, error) {
+		vols, err := volumes.ExtractVolumes(page)
+		if err != nil {
+			glog.Errorf("Failed to extract volumes: %v", err)
+			return false, err
+		}
+
+		for _, v := range vols {
+			glog.V(4).Infof("%s %s %v", v.ID, v.Name, v.Attachments)
+			if v.Name == diskName || strings.Contains(v.ID, diskName) {
+				volume = v
+				return true, nil
+			}
+		}
+
+		// if it reached here then no disk with the given name was found.
+		errmsg := fmt.Sprintf("Unable to find disk: %s in region %s", diskName, rs.region)
+		return false, errors.New(errmsg)
+	})
+	if err != nil {
+		glog.Errorf("Error occured getting volume: %s", diskName)
+	}
+	return volume, err
+}
+
+func (rs *Rackspace) getComputeClient() (*gophercloud.ServiceClient, error) {
+	client, err := rackspace.NewComputeV2(rs.provider, gophercloud.EndpointOpts{
+		Region: rs.region,
+	})
+	if err != nil || client == nil {
+		glog.Errorf("Unable to initialize nova client for region: %s", rs.region)
+	}
+	return client, nil
+}
+
+// Detaches given cinder volume from the compute running kubelet
+func (rs *Rackspace) DetachDisk(partialDiskId string) error {
+	instanceID, err := readInstanceID()
+	if err != nil {
+		return err
+	}
+
+	disk, err := rs.getVolume(partialDiskId)
+	if err != nil {
+		return err
+	}
+
+	compute, err := rs.getComputeClient()
+	if err != nil {
+		return err
+	}
+
+	if len(disk.Attachments) > 0 && instanceID == disk.Attachments[0]["server_id"] {
```

Comment:
Agreed. It appears Rackspace does not support multiattach. So attachments should never be > 1.

> If we want to be defensive/forward-compatible about this, we could bail out with an error when a volume has > 1 attachements (still, currently the Rackspace API does not allow it).

I am a fan of defensive programming, meaning even if we don't expect something to happen we should handle it. So I'd love to see the attachments > 1 cases handled explicitly.